### PR TITLE
P2P: infrastructure enhancing, data conversion, flow infrastructure and IBD pruning sync and Headers download

### PR DESCRIPTION
**Pull request by michaelsutton**
This PR finalizes and considerably enhances and polishes the P2P infrastructure layer. 

It also includes the foundation for implementing all P2P flows, as well as some consensus <-> P2P data conversion logic and a working prototype of IBD -- syncing the pruning proof and downloading headers. To see it in action, have a kaspad golang node running and run:

`
cargo run --release --bin kaspad -- --loglevel=p2p_lib=debug,p2p_flows=debug --connect=localhost:16111 --reset-db
`

### CODE REVIEW

**Review by someone235 — CHANGES_REQUESTED** [review]

File: protocol/flows/src/v5/ibd/flow.rs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        self.start_impl().await
+    }
+}
+
+impl IbdFlow {
+    pub fn new(ctx: FlowContext, router: Arc<Router>, incoming_route: IncomingRoute) -> Self {
+        Self { ctx, router, incoming_route }
+    }
+
+    async fn start_impl(&mut self) -> Result<(), ProtocolError> {
```
[the remark this answers is not in the document]
Comment (michaelsutton) [reply protocol/flows/src/v5/ibd/flow.rs:58]: Yes. Btw shouldn't it be `run`? in go we named it start, but essentially it "runs" the flow

File: protocol/p2p/src/convert/model/trusted.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    /// Returns the trusted set -- a sub-DAG in the anti-future of the pruning point which contains
+    /// all the blocks and ghostdag data needed in order to validate the headers in the future of
+    /// the pruning point
+    pub fn build_trusted_subdag(self, entries: Vec<TrustedDataEntry>) -> Result<Vec<TrustedBlock>, ProtocolError> {
+        let mut blocks = Vec::with_capacity(entries.len());
+        let mut set = BlockHashSet::new();
+        let mut map = BlockHashMap::new();
+
+        for th in self.ghostdag_window.iter() {
```
[the remark this answers is not in the document]
Comment (michaelsutton) [reply protocol/p2p/src/convert/model/trusted.rs:34]: I don't mind being on the safe side here

File: protocol/flows/src/flow_context.rs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+        // Send and receive the ready signal
+        handshake.exchange_ready_messages().await?;
+
+        // Launch all flows. Note we launch only after the ready signal was exchanged
+        for flow in flows {
+            flow.launch();
+        }
+
+        // Note: at this point receivers for handshake subscriptions
```
[the remark this answers is not in the document]
Comment (michaelsutton) [reply protocol/flows/src/flow_context.rs:74]: feature

File: consensus/src/consensus/test_consensus.rs [hunk 4]
```diff
@@ -313,6 +313,9 @@ pub fn create_or_load_existing_db(db_dir: PathBuf) -> Arc<DB> {
 }
 
 pub fn delete_db(db_dir: PathBuf) {
+    if !db_dir.exists() {
```
Comment (someone235) [inline consensus/src/consensus/test_consensus.rs:316]: Move delete_db out of test consensus

**Review by someone235 — APPROVED** [review]

**Pull request merged**